pylink/jlink.py

```diff
@@ -293,6 +309,29 @@ def __del__(self):
             if self.opened():
                 self.close()
 
+    def __enter__(self):
+        """Connects to the J-Link emulator (defaults to USB) using context manager.
+
+        Parameters passed to __init__ are used for open() function.
+
+        Returns:
+          the ``JLink`` instance
+
+        Raises:
+          JLinkException: if fails to open (i.e. if device is unplugged)
+          TypeError: if ``serial_no`` is present, but not ``int`` coercible.
+          AttributeError: if ``serial_no`` and ``ip_addr`` are both ``None``.
+        """
+        if self._open_tunnel is False:
+            self.open(serial_no=self._serial_no, ip_addr=self._ip_addr)
+        elif self._open_tunnel is True:
+            self.open_tunnel(serial_no=self._serial_no)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.__del__()
```

Comment:
Given you're calling `.open()` in `__enter__`, it's probably better that you call `.close()` instead of explicitly calling the deconstructor.

Reply:
Initially I had
```python
    def __exit__(self, exc_type, exc_val, exc_tb):
        try:
            self.close()
        except JLinkException:
            # There is no connected JLink.
            pass
```
to always try to close directly and silently ignore (handle with except) the rare cases when user will manually close before exiting the context manager scope. The reason was that try-except is faster than explicit if-check if no exception is raised.
However for clarity I then anyway changed the code to:
```python
    def __exit__(self, exc_type, exc_val, exc_tb):
        if self.opened():
            self.close()
```
but then I looked at the `__del__` method and thought that the same implementation fits both places. First idea was to move the code from `__del__` to separate private method and call it from `__del__` and `__exit__`, but then saw that while it is true that GC calls `__del__`, calling `__del__` manually does not trigger any additional operations. Although I do agree it is ugly.
Checking for `self._initialized` is not required in `__exit__` as `__init__` will always complete prior to executing `__enter__` and `__exit__`. However I think stopping the eventually enabled SWO is smart idea before closing.
What is your opinion? Shall the check for enabled SWO be moved to `.close()` or better shall this code be moved to e.g. `def _finalize():` and then called from `__exit__` and `__del__`?
Moreover, [`weakref.finalize`](https://docs.python.org/3/library/weakref.html#weakref.finalize) could be used instead of implementing `__del__` (which is usually considered highly avoidable), linked to e.g. `_finalize` method and then called from `__exit__` (and/or when object is garbage collected, while ensuring it is called at least and exactly once).